src/twisted/persisted/dirdbm.py

```diff
@@ -60,120 +65,134 @@ def __init__(self, name):
-        """Read in the contents of a file.
-        
+        """
+        Read in the contents of a file.
+
         Override in subclasses to e.g. provide transparently encrypted dirdbm.
         """
-        with _open(path, "rb") as f:
+        with _open(path.path, "rb") as f:
             s = f.read()
         return s
-    
+
+
     def _writeFile(self, path, data):
-        """Write data to a file.
-        
+        """
+        Write data to a file.
+
         Override in subclasses to e.g. provide transparently encrypted dirdbm.
         """
-        with _open(path, "wb") as f:
+        with _open(path.path, "wb") as f:
             f.write(data)
             f.flush()
-    
+
+
     def __len__(self):
         """
         @return: The number of key/value pairs in this Shelf
         """
-        return len(os.listdir(self.dname))
+        return len(self._dnamePath.listdir())
+
 
     def __setitem__(self, k, v):
         """
         C{dirdbm[k] = v}
         Create or modify a textfile in this directory
 
-        @type k: str
+        @type k: bytes
         @param k: key to set
-        
-        @type v: str
+
+        @type v: bytes
         @param v: value to associate with C{k}
         """
-        assert type(k) == str, "DirDBM key must be a string"
-        assert type(v) == str, "DirDBM value must be a string"
+        assert type(k) == bytes, "DirDBM key must be bytes"
+        assert type(v) == bytes, "DirDBM value must be bytes"
         k = self._encode(k)
-        
-        # we create a new file with extension .new, write the data to it, and
+
+        # We create a new file with extension .new, write the data to it, and
         # if the write succeeds delete the old file and rename the new one.
-        old = os.path.join(self.dname, k)
-        if os.path.exists(old):
-            new = old + ".rpl" # replacement entry
+        old = self._dnamePath.child(k)
+        if old.exists():
+            new = old.siblingExtension(".rpl") # Replacement entry
         else:
-            new = old + ".new" # new entry
+            new = old.siblingExtension(".new") # New entry
         try:
             self._writeFile(new, v)
         except:
-            os.remove(new)
+            new.remove()
             raise
         else:
-            if os.path.exists(old): os.remove(old)
-            os.rename(new, old)
+            if (old.exists()): old.remove()
+            new.moveTo(old)
+
 
     def __getitem__(self, k):
         """
         C{dirdbm[k]}
         Get the contents of a file in this directory as a string.
-        
-        @type k: str
+
+        @type k: bytes
         @param k: key to lookup
-        
+
         @return: The value associated with C{k}
         @raise KeyError: Raised when there is no such key
         """
-        assert type(k) == str, "DirDBM key must be a string"
-        path = os.path.join(self.dname, self._encode(k))
+        assert type(k) == bytes, "DirDBM key must be bytes"
```

Comment:
Do we want to also change this from assert to raising an explicit TypeError? (Ditto for next one)